esrally/utils/modules.py

```diff
@@ -34,78 +36,89 @@ class ComponentLoader:
 
     """
 
-    def __init__(self, root_path, component_entry_point, recurse=True):
+    def __init__(self, root_path: Union[str, Collection[str]], component_entry_point: str, recurse: bool = True):
         """
         Creates a new component loader.
 
-        :param root_path: An absolute path to a directory which contains the component entry point.
+        :param root_path: An absolute path or list of paths to a directory which contains the component entry point.
         :param component_entry_point: The name of the component entry point. A corresponding file with the extension ".py" must exist in the
         ``root_path``.
         :param recurse: Search recursively for modules but ignore modules starting with "_" (Default: ``True``).
         """
-        self.root_path = root_path
+        self.root_path: Collection[str] = root_path if isinstance(root_path, list) else [str(root_path)]
         self.component_entry_point = component_entry_point
         self.recurse = recurse
         self.logger = logging.getLogger(__name__)
 
-    def _modules(self, module_paths, component_name):
+    def _modules(self, module_paths: Collection[str], component_name: str, root_path: str):
         for path in module_paths:
             for filename in os.listdir(path):
                 name, ext = os.path.splitext(filename)
                 if ext.endswith(".py"):
-                    root_relative_path = os.path.join(path, name)[len(self.root_path) + len(os.path.sep) :]
+                    file_absolute_path = os.path.join(path, filename)
+                    root_absolute_path = os.path.join(path, name)
+                    root_relative_path = root_absolute_path[len(root_path) + len(os.path.sep) :]
                     module_name = "%s.%s" % (component_name, root_relative_path.replace(os.path.sep, "."))
-                    yield module_name
+                    yield module_name, file_absolute_path
 
-    def _load_component(self, component_name, module_dirs):
+    def _load_component(self, component_name: str, module_dirs: Collection[str], root_path: str):
         # precondition: A module with this name has to exist provided that the caller has called #can_load() before.
         root_module_name = "%s.%s" % (component_name, self.component_entry_point)
-        for p in self._modules(module_dirs, component_name):
-            self.logger.debug("Loading module [%s]", p)
-            m = importlib.import_module(p)
-            if p == root_module_name:
+        for name, p in self._modules(module_dirs, component_name, root_path):
+            self.logger.debug("Loading module [%s]: %s", name, p)
+            spec = importlib.util.spec_from_file_location(name, p)
+            if spec is None or spec.loader is None:
+                raise exceptions.SystemSetupError(f"Could not load module [{name}]")
+            m = importlib.util.module_from_spec(spec)
+            spec.loader.exec_module(m)
+            if name == root_module_name:
```

Comment:
The change from `importlib.import_module()` to `importlib.util.spec_from_file_location() + importlib.util.module_from_spec()` is to differentiate between 2 modules with the same name but different absolute paths, e.g. `rally-teams/cars/v1/x_pack/base/config.py` and `rally-teams/cars/v1/apm_tracing/config.py`? WDYT about adding a comment with some context?